### ISSUE TITLE
requireDefaultImportShorthand and friends

### ISSUE THREAD

**Issue by alawatthe** [jscs-dev/node-jscs#1689]
I'd like to have a rule, which allows the following 
```js
import _ from "lodash";
```
but disallows the following (equivalent) code:

```js
import {default as _} from "lodash";
```
---

And a similar rule for default exports. Allow this:
```js
export default {
  field1: value1,
  field2: value2
};
```
and disallow this:
```js
let myObject = {
  field1: value1,
  field2: value2
};
export {myObject as default};
```

I would call these rules ```requireDefaultImportShorthand``` and ```requireDefaultExportShorthand```, but I am open to name changes. The dual rules (the ones with disallow) should also be implemented.

Any opinions?

**Comment by markelog** [comment]
Not sure how would you enforce second case, like what if 
```js
let myObject = {
  field1: value1,
  field2: value2
};
// ... lots of code, or no code at all
myObject.field3 = value3;
export {myObject as default};
```
?

**Comment by alawatthe** [comment]
Ok the way I wanted to implemt this (probably this weekend), is:
For
```js
export default {
  field1: value1,
  field2: value2
};
```
the AST looks somewhat like
```js
{
  "type": "ExportDefaultDeclaration",
  "declaration": {
    "type": "ObjectExpression",
    "properties": [/*...*/]
  }
}
```
So prohibiting this with ```disallowDefaultExportShorthand``` is basically only calling ```disallowNodeTypes``` with ```"ExportDefaultDeclaration"```.

---

The case of ```requireDefaultExportShorthand``` is only a little bit more complicated. Requiring the default export shorthand, means dissallowing code like
```js
export {myObject as default};
```
For this snippet the AST looks somewhat like
```js
{
  "type": "ExportNamedDeclaration",
  "declaration": null,
  "specifiers": [
    {
      "type": "ExportSpecifier",
      "local": {
        "type": "Identifier",
        "name": "myObject"
      },
      "exported": {
        "type": "Identifier",
        "name": "default"
      }
    }
  ],
}
```
So, for the ```ExportSpecifier``` we need to disallow the use of ```"default"``` in the exported.name property.
@markelog I hope, this answers your question. The rules for the imports can be handled accordingly.

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
